File: awesome/widgets/wifi_status.lua
local textbox = require("wibox.widget.textbox")

local colors = require("colors")
local underline = require("widgets.underline")
local wifi_status_daemon = require("daemons.wifi_status")

local signal_name = wifi_status_daemon.signal_name

local wifi_status = {}

local function new(args)
    local color_connected = colors.normal.green
    local color_disconnected = colors.normal.red
    local color_error = colors.normal.red

    local color_connected_marker = colors.normal.green
    local color_disconnected_marker = colors.normal.red
    local color_error_marker = colors.normal.red

    local textbox_widget = textbox("")
    local widget = underline(textbox_widget, color_disconnected)

    local function handler(status, name, signal)
        local color = color_disconnected
        local markup = ""

        if status == wifi_status_daemon.status_connected then
            color = color_connected
            markup = string.format("<span foreground='%s'><b>WIFI</b></span>  <span foreground='%s'>%s</span>  %s%%", color_connected_marker, color_connected, name, signal)
        elseif status == wifi_status_daemon.status_disconnected then
            color = color_disconnected
            markup = string.format("<span foreground='%s'><b>WIFI</b></span>", color_disconnected_marker)
        elseif status == wifi_status_daemon.status_error then
            color = color_error
            markup = string.format("<span foreground='%s'><b>WIFI ERROR</b></span>", color_error_marker)
        else
            color = color_error
            markup = string.format("<span foreground='%s'><b>WIFI UNKNOWN STATUS</b></span>", color_error_marker)
        end

        textbox_widget:set_markup(markup)
        widget:set_color(color)
    end

    handler(wifi_status_daemon.status_disconnected, "", 0)

    -- setmetatable(widget, {
    --   __gc = function()
    --     awesome.disconnect_signal(signal_name, handler)
    --   end
    -- })

    awesome.connect_signal(signal_name, handler)

    return widget
end

return setmetatable(wifi_status, { __call = function(_, ...) return new(...) end })

File: awesome/daemons/wifi_status.ts
'use strict';

import watch = require('awful.widget.watch');

export const signal_name = 'daemons::wifi';

export const status_connected = 'connected';
export const status_disconnected = 'disconnected';
export const status_error = 'error';

const home = os.getenv('HOME') || '~';

/** @tupleReturn */
function parse_result(
  stdout: string | null,
  exitcode: number
): [string, string, number, boolean] {
  if (exitcode !== 0) {
    return [status_error, '', 0, true];
  }

  if (typeof stdout !== 'string') {
    return ['', '', 0, false];
  }

  const [name, signal] = stdout.match('connected%s+([^%s]+)%s+([0-9]+)');
  if (signal === null) {
    const [v] = stdout.match('disconnected');
    if (v) {
      return [status_disconnected, '', 0, true];
    }

    return ['', '', 0, false];
  }

  return [status_connected, name, tonumber(signal) || 0, true];
}

watch(home + '/bin/wifistatus', 10, (_w, stdout, _stderr, _r, exitcode) => {
  const [status, name, signal, ok] = parse_result(stdout, exitcode);
  if (!ok) {
    return;
  }

  awesome.emit_signal(signal_name, status, name, signal);
});

export const _private = {
  parse_result
};
